### PR TITLE
Fix ERC20._update

### PR DESCRIPTION
**Pull request by JulissaDantes**
There is a risk with the current implementation of the `_update` function of altering the tokens total supply when sending the zero address as the `from` and `to` parameters. 

If someone does `_update(address(0), address(0), amount)`, this will go into
`if (from == address(0)) { ... }`
that will increase the total supply and mint tokens to address(0), then because of the `else` it will not go into
`else if (to == address(0)) { ... }`
so the total supply amount would remain unbalanced.

#### PR Checklist

- [x] Tests
- [x] Documentation
- [x] Changelog entry

### CODE REVIEW

File: contracts/token/ERC20/ERC20.sol [hunk 1]
```diff
@@ -241,25 +241,20 @@ contract ERC20 is Context, IERC20, IERC20Metadata {
...
+        if (to == address(0)) {
+            _totalSupply -= amount;
         } else {
-            uint256 fromBalance = _balances[from];
-            require(fromBalance >= amount, "ERC20: transfer amount exceeds balance");
             unchecked {
-                _balances[from] = fromBalance - amount;
-                // Overflow not possible: the sum of all balances is capped by totalSupply, and the sum is preserved by
-                // decrementing then incrementing.
+                // Overflow not possible: balance + amount is at most totalSupply + amount, which is checked above.
```
Comment (Amxx) [inline contracts/token/ERC20/ERC20.sol:257]: That comment is wrong, totalSupply+amount is not checked.

A better comment would be balance+amount is at most totalSupply, which we know fits into a uint256

File: contracts/token/ERC20/ERC20.sol [hunk 2]
```diff
@@ -241,25 +241,20 @@ contract ERC20 is Context, IERC20, IERC20Metadata {
...
-            _totalSupply -= amount;
+            require(fromBalance >= amount, "ERC20: transfer amount exceeds balance");
             unchecked {
                 // Overflow not possible: amount <= fromBalance <= totalSupply.
                 _balances[from] = fromBalance - amount;
             }
+        }
+
+        if (to == address(0)) {
+            _totalSupply -= amount;
```
Comment (Amxx) [inline contracts/token/ERC20/ERC20.sol:254]: this should be unchecked

**Comment by Amxx** [comment]
ERC20Snapshot should also be updated. Should that be done in a separate PR?

File: contracts/token/ERC20/ERC20.sol (hunk 2 above)
Comment (JulissaDantes) [reply contracts/token/ERC20/ERC20.sol:254]: why?

**Comment by JulissaDantes** [comment]
> ERC20Snapshot should also be updated. Should that be done in a separate PR?

It will be done here.

File: contracts/token/ERC20/ERC20.sol (hunk 2 above)
Comment (Amxx) [reply contracts/token/ERC20/ERC20.sol:254]: this cannot overflow:
- either the update is from 0 to 0, in which case we are removing what we just added → this is safe
- or we are updating from X to 0 (burn), in which case the amount burned is constrained by the from balance, which we know cannot be bigger than the totalSupply → this is also safe

File: test/token/ERC20/ERC20.behavior.js [hunk 3]
```diff
@@ -178,6 +178,17 @@ function shouldBehaveLikeERC20 (errorPrefix, initialSupply, initialHolder, recip
         );
       });
     });
+
+    describe('when the token owner and the recipient are the zero address', function () {
```
Comment (Amxx) [inline test/token/ERC20/ERC20.behavior.js:182]: This is checking that the `transferFrom` function is not correctly preventing the 0 address.

We need a test of the internal function (`$_update`) to check that doing from 0 to 0 internally does not revert, emits the right event, and doesn't change the totalsupply

**Comment by Amxx** [comment]
We should merge master into next-5.0. 

This will enable hardhat-exposed, removing the needs for mocks.

File: contracts/token/ERC20/ERC20.sol (hunk 2 above)
Comment (frangio) [reply contracts/token/ERC20/ERC20.sol:254]: Yes, in fact this explanation is included above where it says:

```
// Overflow not possible: amount <= fromBalance <= totalSupply.
```

File: test/token/ERC20/ERC20.test.js [hunk 4]
```diff
@@ -287,6 +287,46 @@ contract('ERC20', function (accounts) {
     });
   });
 
+  describe('_update', function () {
+    const amount = initialSupply + 1;
```
Comment (frangio) [inline test/token/ERC20/ERC20.test.js:291]: This seems simpler. I don't see a reason to involve `initialSupply`.

```suggestion
    const amount = 1;
    
```

File: test/token/ERC20/ERC20.test.js [hunk 5]
```diff
@@ -287,6 +287,46 @@ contract('ERC20', function (accounts) {
...
+    it('to is the zero address', async function () {
+      const totalSupply = await this.token.totalSupply();
+      const burnAmount = await this.token.balanceOf(initialHolder);
+      await this.token.update(initialHolder, ZERO_ADDRESS, burnAmount);
+      
+      expect(await this.token.totalSupply()).to.be.bignumber.equal(totalSupply.sub(burnAmount));
+      expect(await this.token.balanceOf(initialHolder)).to.be.bignumber.equal(new BN(0));
+    }); 
+
+    describe('from and to are the zero address', function () {
```
Comment (frangio) [inline test/token/ERC20/ERC20.test.js:310]: It's odd that this is tested differently to the other two cases.

That said, I'm not sure that from and to zero addresses should be a valid function call. It sounds to me like it should revert.

File: contracts/token/ERC20/ERC20.sol (hunk 2 above)
Comment (frangio) [reply contracts/token/ERC20/ERC20.sol:254]: Yes, in fact this explanation is included above where it says:

```
// Overflow not possible: amount <= fromBalance <= totalSupply.
```

File: test/token/ERC20/ERC20.test.js (hunk 5 above)
Comment (Amxx) [reply test/token/ERC20/ERC20.test.js:310]: > That said, I'm not sure that from and to zero addresses should be a valid function call. It sounds to me like it should revert.

IMO this is a no-op, and we always said that no-op should no revert. Also the whole point of this PR is to fix the behavior when this is called.

File: test/token/ERC20/ERC20.test.js [hunk 6]
```diff
@@ -287,6 +287,46 @@ contract('ERC20', function (accounts) {
     });
   });
 
+  describe('_update', function () {
+    const amount = 1;
+    
+    it('from is the zero address', async function () {
```
Comment (frangio) [inline test/token/ERC20/ERC20.test.js:293]: Let's add an assertion for the transfer event in this test.

File: test/token/ERC20/ERC20.test.js [hunk 7]
```diff
@@ -287,6 +287,46 @@ contract('ERC20', function (accounts) {
...
+      const totalSupply = await this.token.totalSupply();
+      const newSupply = totalSupply.add(new BN(amount));
+      await this.token.update(ZERO_ADDRESS, recipient, amount);
+
+      expect(await this.token.totalSupply()).to.be.bignumber.equal(newSupply);
+      expect(await this.token.balanceOf(recipient)).to.be.bignumber.equal(amount);
+    });
+
+    it('to is the zero address', async function () {
```
Comment (frangio) [inline test/token/ERC20/ERC20.test.js:302]: Let's add an assertion for the transfer event in this test.

File: test/token/ERC20/ERC20.test.js [hunk 8]
```diff
@@ -287,6 +287,46 @@ contract('ERC20', function (accounts) {
...
+      const totalSupply = await this.token.totalSupply();
+      const burnAmount = await this.token.balanceOf(initialHolder);
+      await this.token.update(initialHolder, ZERO_ADDRESS, burnAmount);
+
+      expect(await this.token.totalSupply()).to.be.bignumber.equal(totalSupply.sub(burnAmount));
+      expect(await this.token.balanceOf(initialHolder)).to.be.bignumber.equal(new BN(0));
+    });
+
+    describe('from and to are the zero address', function () {
```
Comment (frangio) [inline test/token/ERC20/ERC20.test.js:311]: Please convert this `describe` block into an `it` that looks like the other two cases. I can't see a reason why there should be a separate test for the event.

**Review by frangio — CHANGES_REQUESTED** [review]

File: test/token/ERC20/ERC20.test.js (hunk 6 above)
Comment (JulissaDantes) [reply test/token/ERC20/ERC20.test.js:293]: Fixed in last commit.

File: test/token/ERC20/ERC20.test.js (hunk 7 above)
Comment (JulissaDantes) [reply test/token/ERC20/ERC20.test.js:302]: Fixed in last commit.

File: test/token/ERC20/ERC20.test.js (hunk 8 above)
Comment (JulissaDantes) [reply test/token/ERC20/ERC20.test.js]: Fixed in last commit.

File: contracts/token/ERC20/ERC20.sol (hunk 2 above)
Comment (frangio) [reply contracts/token/ERC20/ERC20.sol:254]: Still need to add unchecked to this line.

**Comment by frangio** [comment]
Really confused by why the linter is still failing. It doesn't show any errors locally.

**Comment by Amxx** [comment]
> Really confused by why the linter is still failing. It doesn't show any errors locally.

same here

**Comment by frangio** [comment]
I was somehow able to fix linting by running the latest version of Prettier with the new configuration.

**Comment by Amxx** [comment]
> I was somehow able to fix linting by running the latest version of Prettier with the new configuration.

is it possible that github action as a cache with the old version ?

**Comment by frangio** [comment]
I tried deleting the caches and it didn't make a difference.

**Comment by Amxx** [comment]
Note that linting of the mocks is irrelevant as they will be deleted when we merge master.

**Review by frangio — APPROVED** [review]

**Pull request merged**